### PR TITLE
Parse Protobuf and Avro schema

### PR DESCRIPTION
**Pull request by usefulalgorithm**
### 🤔 Why?

Parse protobuf and avro schema in Kafka topics so that they can be properly ingested & displayed.

### 🤓 What?

- Added logic to parse protobuf and avro schemas.
- Very basic unit test.

### 🧪 Tested?

- Unit test.

### ☑️ Checks

- [x] My PR contains actual code changes, and I have updated the version number in `pyproject.toml`.

### CODE REVIEW

**Comment by usefulalgorithm** [comment]
I'll add more solid tests and make sure UI is able to display the parsed schema fields. Meanwhile the main logic should be ready for review.

**Comment by usefulalgorithm** [comment]
@kiritbasu This is the PR I was talking about during standup. Basically I am expecting the below conditions to be met when running the connector:
- there is a topic
- have a subject in the schema registry that corresponds to the topic
  - in confluent there are different strategies to map topics to schemas. I did those as well.
- a subject corresponds to a schema
- the schema is well formed, i.e. no garbage schema

If these conditions all hold then I think this would work, but I don't know if all of these assumptions make sense for normal users.

File: tests/kafka/schema_parsers/test_avro_parser.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        SchemaField(
+            description="application messages",
+            field_name="myRecord",
+            field_path="myRecord",
+            native_type="RECORD",
+            subfields=[
+                SchemaField(
+                    field_name="requestResponse",
+                    field_path="myRecord.requestResponse",
+                    native_type="UNION<record,record>",  # FIXME Where are the subfields?
```
Comment (usefulalgorithm) [inline tests/kafka/schema_parsers/test_avro_parser.py:272]: Note to self: fix this

File: metaphor/kafka/schema_parsers/avro_parser.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            schema_field.subfields = [subfields]
+        return schema_field
+
+    def _parse_union_children(
+        self,
+        parent: Optional[avroschema.Schema],
+        union_field: UnionSchema,
+        cur_path: str,
+    ) -> Tuple[str, Optional[SchemaField]]:
+        # FIXME the logic here is incorrect
```
Comment (usefulalgorithm) [inline metaphor/kafka/schema_parsers/avro_parser.py:126]: Note to self: fix this

File: pyproject.toml [hunk 3]
```diff
@@ -116,6 +120,7 @@ types-python-dateutil = "^2.8.19.12"
 types-pytz = "^2023.3.0.0"
 types-PyYAML = "^6.0.4"
 types-requests = "^2.28.11"
+types-protobuf = "^4.24.0.4"
```
Comment (mars-lan) [inline pyproject.toml:123]: Alphabetical order.

File: tests/kafka/schema_parsers/test_avro_parser.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
+from metaphor.kafka.schema_parsers.avro_parser import AvroParser
+from metaphor.models.metadata_change_event import SchemaField
+
+
+def test_parse_schema() -> None:
+    raw_schema = """
+{
```
Comment (mars-lan) [inline tests/kafka/schema_parsers/test_avro_parser.py:7]: Let's define this in a `.json` file to improve readability and apply auto-formating.

**Review by mars-lan — COMMENTED** [review]
LGTM from a high-level. @elic-eon pls give it a deeper review.

File: metaphor/kafka/schema_parsers/protobuf_parser.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
+import importlib
+import shutil
+import subprocess
+import sys
+import tempfile
+from enum import Enum
+from pathlib import Path
+from typing import List, Optional
+
+from google.protobuf.descriptor import FileDescriptor  # type: ignore
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/protobuf_parser.py:10]: Do we need to ignore `mypy` checking here?

File: metaphor/kafka/schema_parsers/protobuf_parser.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        message = importlib.import_module(module_name)
+        top_level_message = self._find_top_level_message(message)
+        if not top_level_message:
+            raise ValueError("Cannot determine top level message")
+        class_ = getattr(message, top_level_message)
+        instance = class_()
+        return instance
+
+    def _find_top_level_message(self, message) -> Optional[str]:
+        # Phew that's a lot of reflection
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/protobuf_parser.py:96]: Wanna drop this?

File: metaphor/kafka/schema_parsers/protobuf_parser.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+        file_desc: FileDescriptor = message.DESCRIPTOR
+
+        # Collect message types that are referenced in other messages
+        referenced_message_types = set()
+        for message in file_desc.message_types_by_name.values():
+            for field in message.fields:
+                if (
+                    field.type == 11
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/protobuf_parser.py:105]: Use the enum value here?

File: metaphor/kafka/schema_parsers/protobuf_parser.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                return message.name
+
+        return None  # If no top-level message is found
+
+    def _parse_protobuf_fields(self, fields, cur_path: str) -> List[SchemaField]:
+        schema_fields = []
+        for field in fields:
+            field_path = ".".join([cur_path, field.name])
+            if cur_path == "":
+                field_path = field_path[1:]
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/protobuf_parser.py:123]: Not sure if there is a simpler way to drop the leading `sep` if the first array element is empty.
```suggestion
            field_path = ".".join([cur_path, field.name]) if cur_path else field.name
```

File: metaphor/kafka/schema_parsers/protobuf_parser.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+            schema_field = SchemaField(
+                field_name=field.name,
+                field_path=field_path,
+                native_type=ProtobufDataTypes(field.type).name,
+                subfields=(
+                    self._parse_protobuf_fields(field.message_type.fields, field_path)
+                    if field.type == 11
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/protobuf_parser.py:131]: ditto?

File: metaphor/kafka/schema_parsers/protobuf_parser.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            )
+            schema_fields.append(schema_field)
+
+        return schema_fields
+
+    def run_parser(self, raw_schema: str, schema_name: str) -> List[SchemaField]:
+        file_path = self._create_proto_file(raw_schema, schema_name)
+        instance = self._get_protobuf_python_object(file_path, schema_name)
+        schema_fields = self._parse_protobuf_fields(
+            instance.DESCRIPTOR.fields, cur_path=""
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/protobuf_parser.py:143]: Is `cur_path` always an empty string?

File: metaphor/kafka/schema_parsers/avro_parser.py [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+    def _safe_get_doc(schema: avroschema.Schema) -> Optional[str]:
+        return (
+            str(schema.get_prop("doc")) if schema.get_prop("doc") is not None else None
+        )
+
+    @staticmethod
+    def _get_field_path(cur_path: str, field_name: str) -> str:
+        # In our case here `cur_path` will never be empty, so no need for
+        # edge condition handling.
+        return ".".join([cur_path, field_name])
```
Comment (elic-eon) [inline metaphor/kafka/schema_parsers/avro_parser.py:33]: Consider moving this logic to utils to handle the edge case, no matter which parser you are using?

File: metaphor/kafka/schema_parsers/protobuf_parser.py (hunk 5 above)
Comment (usefulalgorithm) [reply metaphor/kafka/schema_parsers/protobuf_parser.py:10]: yes, even though I installed types-protobuf my IDE is still complaining about things not found.

File: metaphor/kafka/schema_parsers/protobuf_parser.py (hunk 6 above)
Comment (usefulalgorithm) [reply metaphor/kafka/schema_parsers/protobuf_parser.py:96]: I'll change it into something more informative

File: metaphor/kafka/schema_parsers/avro_parser.py (hunk 11 above)
Comment (usefulalgorithm) [reply metaphor/kafka/schema_parsers/avro_parser.py:33]: sf

File: metaphor/kafka/schema_parsers/protobuf_parser.py (hunk 10 above)
Comment (usefulalgorithm) [reply metaphor/kafka/schema_parsers/protobuf_parser.py:141]: yes, for protobuf I don't think we want the message type name to be in field path:

```protobuf
message Foo {
    string bar = 1;
}
```

then it should be
```python
SchemaField(
    field_name="bar",
    field_path="bar",
    ...
)
```

Comment (elic-eon) [reply metaphor/kafka/schema_parsers/protobuf_parser.py:147]: Got it. I missed the L134.

**Review by elic-eon — APPROVED** [review]
Great work!

**Comment by kiritbasu** [comment]
@usefulalgorithm couple of general points : 
- when reading from schema registry its useful to know
  - Schema version
  - id
  - FQN i.e. namespace 
 - (nice to have) hyperlinks back to where schema registry is hosted 
- (if you haven't already seen it) might be good to quickly scan through https://github.com/confluentinc/schema-registry/tree/master they might have comprehensive test cases for avro/protobuf..  I don't think you need to replicate their full set of tests but just look though if anything stands out.

**Pull request merged**